deepinv/physics/generator/mri.py

```diff
@@ -1,83 +1,244 @@
+    ::
+
+        generator = GaussianMaskGenerator((2, 8, 64, 64), acceleration=8, center_fraction=0.04) # C, T, H, W
+        params = generator.step(batch_size=1)
+        mask = params["mask"]
+        mask.shape # 1, 2, 8, 64, 64
+
+    """
+
+    def get_pdf(self) -> torch.Tensor:
+        """Create one-dimensional Gaussian probability density function across columns, ignoring any fixed sampling columns.
+
+        Add uniform distribution so that the probability of sampling high-frequency lines is non-zero.
+
+        :return torch.Tensor: unnormalised 1D vector representing pdf evaluated across mask columns.
+        """
+        normal_pdf = lambda length, sensitivity: np.exp(
+            -sensitivity * (np.arange(length) - length / 2) ** 2
         )
-        mask[:, :, center_line_indices] = 1
+        pdf = normal_pdf(self.W, 0.5 / (self.W / 10.0) ** 2)
+        lmda = self.W / (2.0 * self.acc)
 
-        for i in range(batch_size):
-            random_line_indices = np.random.choice(
-                H, size=(num_lines_side // 2,), replace=False
-            )
-            mask[i, :, random_line_indices] = 1
+        pdf += lmda * 1.0 / self.W
+
+        return pdf
+
+
+class UniformMaskGenerator(BaseMaskGenerator):
+    """Generator for MRI Cartesian acceleration masks using uniform (equispaced) non-random undersampling.
+
+    Generate a mask of vertical lines for MRI acceleration with fixed sampling in low frequencies (center of k-space) and equispaced undersampling in the high frequencies.
+
+    The number of lines selected with equal spacing are at a proportion that reaches the desired acceleration rate taking into consideration the number of low-freq lines, so that the total number of lines is (N / acceleration).
+
+    Supports k-t sampling, where the uniform mask is sheared across time.
 
-        return {"mask": torch.cat([mask.float().unsqueeze(1)] * C, dim=1)}
+    The mask is repeated across channels and randomly varies across batch dimension. Based off fastMRI code https://github.com/facebookresearch/fastMRI
+
+    For parameter descriptions see :class:`deepinv.physics.generator.mri.BaseMaskGenerator`
+
+    Example:
+
+    ::
+
+        generator = UniformMaskGenerator((2, 8, 64, 64), acceleration=8, center_fraction=0.04) # C, T, H, W
+        params = generator.step(batch_size=1)
+        mask = params["mask"]
+        mask.shape # 1, 2, 8, 64, 64
+
+    """
+
+    def sample_mask(self, mask: torch.Tensor) -> torch.Tensor:
+        pad = (self.W - self.n_center + 1) // 2
+        mask[:, :, :, :, pad : pad + self.n_center] = 1
+
+        # determine acceleration rate by adjusting for the number of low frequencies
+        adjusted_accel = (self.acc * (self.n_center - self.W)) / (
+            self.n_center * self.acc - self.W
+        )
+        offset = self.rng.integers(0, round(adjusted_accel))
+
+        for b in range(mask.shape[0]):
+            for t in range(mask.shape[2]):
+                accel_samples = (
```

Comment:
I'm also worried here about speed

Reply:
I don't think doing `np.arange`s per batch is a big problem - it should be quite quick?

Reply:
ok